fish-rust/src/fd_monitor.rs

```diff
@@ -0,0 +1,573 @@
+        /// [`select()`](libc::select).
+        ///
+        /// This is like a binary semaphore. A call to [`post()`](FdEventSignaller::post) will
+        /// signal an event, making the fd readable.  Multiple calls to `post()` may be coalesced.
+        /// On Linux this uses [`eventfd()`](libc::eventfd), on other systems this uses a pipe.
+        /// [`try_consume()`](FdEventSignaller::try_consume) may be used to consume the event.
+        /// Importantly this is async signal safe. Of course it is `CLO_EXEC` as well.
+        #[rust_name = "FdEventSignaller"]
+        type fd_event_signaller_t = crate::ffi::fd_event_signaller_t;
+        // /// Mark that an event has been received.
+        // /// This may be coalesced. This retries on `EINTR`.
+        // /// TODO: This needs to be `&self` when ported to rust, otherwise it can't be used from two
+        // /// different threads.
+        // fn post(&self);
+        // /// Returns the fd to read from for notification.
+        // fn read_fd(&self) -> i32;
+        // /// If an event is signalled, consume it; otherwise return.
+        // /// This does not block. This retries on `EINTR`.
+        // fn try_consume(&self) -> bool;
+        #[rust_name = "new_fd_event_signaller"]
+        fn ffi_new_fd_event_signaller_t() -> UniquePtr<FdEventSignaller>;
+    }
+    // unsafe extern "C++" {
+    //     include!("fds.h");
+    //
+    //     /// A helper class for managing and automatically closing a file descriptor.
+    //     #[rust_name = "AutoCloseFd"]
+    //     type autoclose_fd_t; // = crate::ffi::autoclose_fd_t;
+    //     /// Closes the fd if not already closed
+    //     fn close(self: Pin<&mut AutoCloseFd>);
+    //     /// Returns the internal/raw fd.
+    //     fn fd(&self) -> i32;
+    //     /// Returns the fd, transferring ownership to the caller
+    //     fn acquire(self: Pin<&mut AutoCloseFd>) -> i32;
+    //     /// Resets to a new fd, taking ownership.
+    //     fn reset(self: Pin<&mut AutoCloseFd>, fd: i32);
+    //     /// Returns `true` if this `AutoCloseFd` instance has a valid fd (`>= 0`).
+    //     #[rust_name = "is_valid"]
+    //     fn valid(&self) -> bool;
+    // }
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_id_t"]
+        type FdMonitorItemId;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_item_t"]
+        type FdMonitorItem;
+
+        #[cxx_name = "make_fd_monitor_item_t"]
+        fn new_fd_monitor_item_ffi(fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> Box<FdMonitorItem>;
+    }
+
+    extern "Rust" {
+        #[cxx_name = "fd_monitor_t"]
+        type FdMonitor;
+
+        #[cxx_name = "make_fd_monitor_t"]
+        fn new_fd_monitor_ffi() -> Box<FdMonitor>;
+
+        #[cxx_name = "add_item"]
+        fn add_item_ffi(&mut self, fd: i32, timeout_usecs: u64, callback: *const c_void, param: *const c_void) -> u64;
```

Comment:
Might have been a bit less confusing for me if we kept the `add()` function and add a separate `new_fd_monitor_item()` function.

Reply:
I think I mentioned somewhere the reason I didn't do that is because each fd_monitor_item would need to be dynamically allocated. We generate dozens of them per pipeline so it seemed stupid to force the ffi caller to allocate a Box only for the rust side of the ffi code to unwrap the box on the next line (when the call to add happens).